### PR TITLE
Openbmc rflash activate (PNOR activation only)

### PR DESCRIPTION
**Pull request by gurevichmark**
This pull request implements the "activate" action for openbmc `rflash` command. Issue #3273 

Example output:

1. Two host update files have been uploaded and `rflash --list` shows:
```
[root@stratton01 xcat]# rflash p9euh02 -l
p9euh02: ID       Purpose State      Version
p9euh02: -------------------------------------------------------
p9euh02: 2a1022fe Host    Ready      IBM-witherspoon-sequoia-ibm-OP9_v1.17_1.67
p9euh02: f3b29aa8 BMC     Active     0.1.0
p9euh02: 32f129cd Host    Ready      IBM-witherspoon-redbud-ibm-OP9_v1.17_1.68
p9euh02:
[root@stratton01 xcat]#
```

2. Activate one of the updates in "Ready" state:
```
[root@stratton01 xcat]# rflash p9euh02 2a1022fe -a
p9euh02: rflash started, please wait...
p9euh02: Activating firmware update. 10%
p9euh02: Activating firmware update. 60%
p9euh02: Activating firmware update. 60%
p9euh02: Activating firmware update. 60%
p9euh02: Activating firmware update. 60%
p9euh02: Firmware update successfully activated
[root@stratton01 xcat]# 
```

3. List will show:
```
[root@stratton01 xcat]# rflash p9euh02 -l
p9euh02: ID       Purpose State      Version
p9euh02: -------------------------------------------------------
p9euh02: 2a1022fe Host    Active(*)  IBM-witherspoon-sequoia-ibm-OP9_v1.17_1.67
p9euh02: f3b29aa8 BMC     Active     0.1.0
p9euh02: 32f129cd Host    Ready      IBM-witherspoon-redbud-ibm-OP9_v1.17_1.68
p9euh02:
```

4. Activate another update in "Ready" state:
```
[root@stratton01 xcat]# rflash p9euh02 32f129cd -a
[OpenBMC development support] Using this version of xCAT, ensure firmware level is at v1.99.6-0-r1, or higher.
p9euh02: rflash started, please wait...
p9euh02: Activating firmware update. 10%
p9euh02: Activating firmware update. 60%
p9euh02: Activating firmware update. 60%
p9euh02: Activating firmware update. 60%
p9euh02: Activating firmware update. 60%
p9euh02: Firmware update successfully activated
[root@stratton01 xcat]#
```

5. List will show:
```
[root@stratton01 xcat]# rflash p9euh02 -l
p9euh02: ID       Purpose State      Version
p9euh02: -------------------------------------------------------
p9euh02: 2a1022fe Host    Active     IBM-witherspoon-sequoia-ibm-OP9_v1.17_1.67
p9euh02: f3b29aa8 BMC     Active     0.1.0
p9euh02: 32f129cd Host    Active(*)  IBM-witherspoon-redbud-ibm-OP9_v1.17_1.68
p9euh02:
[root@stratton01 xcat]#
```

6. Activate update already in the "Active" state:
```
[root@stratton01 xcat]# rflash p9euh02 2a1022fe -a
p9euh02: rflash started, please wait...
p9euh02: Firmware update successfully activated
[root@stratton01 xcat]#
```

7. List will show:
```
[root@stratton01 xcat]# rflash p9euh02 -l
p9euh02: ID       Purpose State      Version
p9euh02: -------------------------------------------------------
p9euh02: 2a1022fe Host    Active(*)  IBM-witherspoon-sequoia-ibm-OP9_v1.17_1.67
p9euh02: f3b29aa8 BMC     Active     0.1.0
p9euh02: 32f129cd Host    Active     IBM-witherspoon-redbud-ibm-OP9_v1.17_1.68
p9euh02:
[root@stratton01 xcat]#
```

### CODE REVIEW

**Comment by whowutwut** [comment]
@hu-weihua  Looks like the last merge commit made here failed the Travis CI with " PR FORMAT ERROR : Miss description."   What description is it checking for?  I think this is a default merge commit description from Git, should we accept it?  ==>  `Merge branch 'master' into openbmc_rflash_activate`

**Comment by whowutwut** [comment]
@gurevichmark We need to add some changes for the man page and command line usage for this right? I didn't see any description here:  http://xcat-docs.readthedocs.io/en/latest/guides/admin-guides/references/man1/rflash.1.html

**Comment by gurevichmark** [comment]
I think the Travis error happened because I created the pull request with a blank initial description to just commit the changes. I later went back and added the description to include the output from the command. But it looks like the check has already be done. Perhaps there is a way to rerun it ?

**Comment by gurevichmark** [comment]
For man page update, I have a separate issue #3281 
My plan was to implement all the `rflash` functionality first, then update man page and usage with the final set of flags and options.

**Comment by whowutwut** [comment]
@gurevichmark Regarding the man page issue, do you think we should update the man pages with the planned externals, then if things change after development, we can update it again... otherwise if we merged this, the user has no idea how to use the rflash.   It would also be good to get feedback on the usage

File: xCAT-server/lib/xcat/plugins/openbmc.pm [hunk 1]
```diff
@@ -562,15 +585,20 @@ sub parse_args {
...
         }
     } elsif ($command eq "rflash") {
-        #
-        # disable function until fully tested
-        #
-        $check = unsupported($callback); if (ref($check) eq "ARRAY") { return $check; }
         my $filename_passed = 0;
+        my $updateid_passed = 0;
         foreach my $opt (@$extrargs) {
+            print "Examening $opt \n";
```
Comment (whowutwut) [inline xCAT-server/lib/xcat/plugins/openbmc.pm]: Spelled wrong... does rflash work for Hierarchy?  If so ,we might want to put the debug messages into the xCAT printer...

File: xCAT-server/lib/xcat/plugins/openbmc.pm [hunk 2]
```diff
@@ -562,15 +585,20 @@ sub parse_args {
...
             # Only files ending on .tar are allowed
             if ($opt =~ /.*\.tar$/i) {
                 $filename_passed = 1;
+                print "Opt matches filename ending on .tar \n";
```
Comment (whowutwut) [inline xCAT-server/lib/xcat/plugins/openbmc.pm]: What is "Opt"  This is the success case right? Do we need to have this printed out?

File: xCAT-server/lib/xcat/plugins/openbmc.pm [hunk 3]
```diff
@@ -1792,6 +1859,54 @@ sub rflash_response {
...
+                return;
+            }
+            else {
+                # Activation state of active and priority of non 0 - need to just set priority to 0 to activate
+                print "Update is already active, just need to set priority to 0";
+                $next_status{ $node_info{$node}{cur_status} } = "RFLASH_SET_PRIORITY_REQUEST";
+            }
+        }
+
+        #if ($activation_state eq "xyz.openbmc_project.Software.Activation.Activations.Activating") {
```
Comment (whowutwut) [inline xCAT-server/lib/xcat/plugins/openbmc.pm]: Should we remove this line?

File: xCAT-server/lib/xcat/plugins/openbmc.pm [hunk 4]
```diff
@@ -1792,6 +1859,54 @@ sub rflash_response {
...
+        if ($activation_state =~ /Software.Activation.Activations.Activating/) {
+            # Activation still going, sleep for a bit, then print the progress value
+            sleep(15);
```
Comment (whowutwut) [inline xCAT-server/lib/xcat/plugins/openbmc.pm]: Are there planned callbacks?

**Comment by whowutwut** [comment]
@gurevichmark I  ran the CI testing, take a look at the xCAT BOT comment , it's been updated with a different error.

File: xCAT-server/lib/xcat/plugins/openbmc.pm (hunk 1 above)
Comment (gurevichmark) [inline xCAT-server/lib/xcat/plugins/openbmc.pm]: Will remove left over debug pring

File: xCAT-server/lib/xcat/plugins/openbmc.pm (hunk 3 above)
Comment (gurevichmark) [inline xCAT-server/lib/xcat/plugins/openbmc.pm]: Will remove this comment

File: xCAT-server/lib/xcat/plugins/openbmc.pm (hunk 4 above)
Comment (gurevichmark) [inline xCAT-server/lib/xcat/plugins/openbmc.pm]: Yes. Currently we land here on each iteration of monitoring PNOR activation.
Not sure yet how BMC activation will behave.

**Comment by gurevichmark** [comment]
1. Will update the man page to reflect current implementation.
2. Title of pull request updated.
3. Yes. Good catch, will update option parsing.

**Comment by gurevichmark** [comment]
Replaced by #3586 (merge problems)

**Pull request closed without merging**